File: src/test/java/br/com/blz/testjava/adapter_out/repository/product/ProductRepositoryTest.java
package br.com.blz.testjava.adapter_out.repository.product;

import br.com.blz.testjava.adapter_in.exception.custom_exception.DuplicateRecordException;
import br.com.blz.testjava.adapter_in.exception.custom_exception.ResultNotFoundException;
import br.com.blz.testjava.adapter_out.entity.InventoryEntity;
import br.com.blz.testjava.adapter_out.entity.ProductEntity;
import br.com.blz.testjava.adapter_out.entity.WareHousesEntity;
import br.com.blz.testjava.adapter_out.mapper.IProductOutMapper;
import br.com.blz.testjava.application.domain.Inventory;
import br.com.blz.testjava.application.domain.Product;
import br.com.blz.testjava.application.domain.WareHouses;
import org.junit.jupiter.api.Assertions;
import org.junit.jupiter.api.DisplayName;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.extension.ExtendWith;
import org.mockito.InjectMocks;
import org.mockito.Mock;
import org.mockito.Mockito;
import org.mockito.junit.jupiter.MockitoExtension;

import java.util.ArrayList;
import java.util.List;

@ExtendWith(MockitoExtension.class)
class ProductRepositoryTest {

    @InjectMocks
    private ProductRepository productRepository;

    @Mock
    private IProductOutMapper productMapper;


    @Test
    @DisplayName("test save product in mockdb")
    void saveProductTest(){
        ProductEntity productEntity = createProductEntity();
        Product product = createProduct();

        Mockito.when(productMapper.toEntity(product)).thenReturn(productEntity);
        Mockito.when(productMapper.toDomain(productEntity)).thenReturn(product);

        Product productSalve = productRepository.saveProduct(product);

        Assertions.assertEquals(product.getSku(), productSalve.getSku());
    }

    @Test
    @DisplayName("test to save duplicate product in mockdb")
    void saveProductDuplicateExceptionTest(){
        ProductEntity productEntity = createProductEntity();
        Product product = createProduct();

        Mockito.when(productMapper.toEntity(product)).thenReturn(productEntity);
        Mockito.when(productMapper.toDomain(productEntity)).thenReturn(product);

        Product productSalve = productRepository.saveProduct(product);

        Assertions.assertEquals(product.getSku(), productSalve.getSku());

        Assertions.assertThrows(DuplicateRecordException.class, () -> productRepository.saveProduct(product));
    }

    @Test
    @DisplayName("product update test in mockdb")
    void updateProductTest(){

        ProductEntity productEntity = createProductEntity();
        Product product = createProduct();

        Mockito.when(productMapper.toEntity(product)).thenReturn(productEntity);
        Mockito.when(productMapper.toDomain(productEntity)).thenReturn(product);

        Product productSalve = productRepository.saveProduct(product);
        Assertions.assertEquals(product.getSku(), productSalve.getSku());


        ProductEntity updateProductEntity = updateProductEntity();
        Product updateProduct = updateProduct();

        Mockito.when(productMapper.toEntity(updateProduct)).thenReturn(updateProductEntity);
        Mockito.when(productMapper.toDomain(updateProductEntity)).thenReturn(updateProduct);


        Product productSalveUpdate = productRepository.updateProduct(updateProduct);
        Assertions.assertEquals(product.getSku(), productSalveUpdate.getSku());
        Assertions.assertEquals(2, productSalveUpdate.getInventory().getWareHousesList().size());

    }

    @Test
    @DisplayName("sku search test in mockdb")
    void findProductTest(){

        ProductEntity productEntity = createProductEntity();
        Product product = createProduct();

        Mockito.when(productMapper.toEntity(product)).thenReturn(productEntity);
        Mockito.when(productMapper.toDomain(productEntity)).thenReturn(product);

        Product productSalve = productRepository.saveProduct(product);
        Assertions.assertEquals(product.getSku(), productSalve.getSku());


        Long sku = 43264L;
        Product produtoFind = productRepository.findProduct(sku);
        Assertions.assertEquals(sku, produtoFind.getSku());

    }

    @Test
    @DisplayName("sku search test for a product, not found in mockdb")
    void findProductNotFoundExceptionTest(){
        Long sku = 43264L;
        Assertions.assertThrows(ResultNotFoundException.class,
                               ()-> productRepository.findProduct(sku));
    }

    @Test
    @DisplayName("test to delete a product by sku in mockdb")
    void deleteProductTest(){

        ProductEntity productEntity = createProductEntity();
        Product product = createProduct();

        Mockito.when(productMapper.toEntity(product)).thenReturn(productEntity);
        Mockito.when(productMapper.toDomain(productEntity)).thenReturn(product);

        Product productSalve = productRepository.saveProduct(product);
        Assertions.assertEquals(product.getSku(), productSalve.getSku());

        Long sku = 43264L;
        productRepository.deleteProduct(sku);
    }

    @Test
    @DisplayName("teste de deletar  um produto pelo sku, não encontrado em mockdb")
    void deleteProductNotFoudExceptionTest(){
        Long sku = 43264L;
        Assertions.assertThrows(ResultNotFoundException.class,
            ()-> productRepository.deleteProduct(sku));
    }


    private Product createProduct() {

        List<WareHouses> wareHousesList = new ArrayList<>();
        wareHousesList.add(
            WareHouses.builder()
                .locality("SP")
                .quantity(12)
                .type("ECOMMERCE")
                .build()
        );

        return Product.builder()
            .sku(43264L)
            .name("L'Oréal Professionnel Expert Absolut Repair Cortex Lipidium - Máscara de Reconstrução 500g")
            .inventory(
                Inventory.builder()
                    .wareHousesList(wareHousesList)
                    .build())
            .build();
    }

    private Product updateProduct() {

        Product product = createProduct();

        product
            .getInventory()
            .getWareHousesList()
            .add(
                WareHouses.builder()
                    .locality("MOEMA")
                    .quantity(3)
                    .type("PHYSICAL_STORE")
                    .build()
            );

        return product;
    }

    private ProductEntity createProductEntity() {

        List<WareHousesEntity> wareHousesEntityList = new ArrayList<>();
        wareHousesEntityList.add(
            WareHousesEntity.builder()
                .locality("SP")
                .quantity(12)
                .type("ECOMMERCE")
                .build()
        );

        return ProductEntity.builder()
            .sku(43264L)
            .name("L'Oréal Professionnel Expert Absolut Repair Cortex Lipidium - Máscara de Reconstrução 500g")
            .inventoryEntity(
                InventoryEntity.builder()
                    .wareHousesEntityList(wareHousesEntityList)
                    .build())
            .build();
    }

    private ProductEntity updateProductEntity() {

        ProductEntity productEntity = createProductEntity();

        productEntity
            .getInventoryEntity()
            .getWareHousesEntityList()
            .add(
                WareHousesEntity.builder()
                    .locality("MOEMA")
                    .quantity(3)
                    .type("PHYSICAL_STORE")
                    .build()
            );

        return productEntity;
    }
}

File: src/test/kotlin/br/com/blz/testjava/domain/repositories/ProductRepositoryTest.kt
package br.com.blz.testjava.domain.repositories

import br.com.blz.testjava.application.exception.NotFoundException
import br.com.blz.testjava.domain.entities.Inventory
import br.com.blz.testjava.domain.entities.Product
import br.com.blz.testjava.domain.entities.Warehouse
import br.com.blz.testjava.domain.entities.WarehouseType
import org.junit.jupiter.api.Assertions.*
import org.junit.jupiter.api.Test
import org.junit.jupiter.api.function.Executable

internal class ProductRepositoryTest {

  @Test
  fun `Must find all products`() {
    val productRepository: ProductRepository = ProductRepository()
    val productsCount = productRepository.findAll().size
    val warehouse = Warehouse("PR", 100, WarehouseType.ECOMMERCE)
    val inventory = Inventory(mutableListOf(warehouse))
    productRepository.save(Product(name = "Product 1", inventory = inventory))
    productRepository.save(Product(name = "Product 2", inventory = inventory))
    productRepository.save(Product(name = "Product 3", inventory = inventory))

    assertEquals(productsCount + 3, productRepository.findAll().size)
  }

  @Test
  fun `Must find a Product by a sku`() {
    val productRepository: ProductRepository = ProductRepository()
    val warehouse = Warehouse("PR", 100, WarehouseType.ECOMMERCE)
    val inventory = Inventory(mutableListOf(warehouse))
    val product = productRepository.save(Product(name = "Product 4", inventory = inventory))

    assertNotNull(productRepository.find(product.sku!!))
  }

  @Test
  fun `Try to find a Product with an incorrect sku must return null`() {
    val productRepository: ProductRepository = ProductRepository()

    assertNull(productRepository.find(112233))
  }

  @Test
  fun `Must insert a product and return one with and sku`() {
    val productRepository: ProductRepository = ProductRepository()
    val inventory = Inventory(mutableListOf(Warehouse("PR", 100, WarehouseType.ECOMMERCE)))
    val product = Product(name = "Product 5", inventory = inventory)
    val saveProduct = productRepository.save(product)

    assertNotNull(saveProduct.sku)
  }

  @Test
  fun `Must update the name of a product`() {
    val productRepository: ProductRepository = ProductRepository()
    val inventory = Inventory(mutableListOf(Warehouse("PR", 100, WarehouseType.ECOMMERCE)))
    val product = productRepository.save(Product(name = "Product 6", inventory = inventory))

    product.name += "_"
    productRepository.update(product)

    val updateProduct = productRepository.find(product.sku!!)
    assertEquals(product.name, updateProduct?.name)
  }


  @Test
  fun `Must update the warehouses of an inventory`() {
    val productRepository: ProductRepository = ProductRepository()
    val warehouse = Warehouse("PR", 100, WarehouseType.ECOMMERCE)
    val inventory = Inventory(mutableListOf(warehouse))
    val product = productRepository.save(Product(name = "Product 7", inventory = inventory))

    val newInventory = Inventory(
      mutableListOf(
        warehouse.apply { quantity = 80 },
        Warehouse("PR", 105, WarehouseType.PHYSICAL_STORE)
      )
    )
    product.inventory = newInventory
    productRepository.update(product)

    val updateProduct = productRepository.find(product.sku!!)!!

    assertEquals(product.inventory.warehouses, updateProduct.inventory.warehouses)
  }

  @Test
  fun `Must delete a product`() {
    val productRepository: ProductRepository = ProductRepository()
    val warehouse = Warehouse("PR", 100, WarehouseType.ECOMMERCE)
    val inventory = Inventory(mutableListOf(warehouse))
    val product = productRepository.save(Product(name = "Product 8", inventory = inventory))

    assertNotNull(productRepository.find(product.sku!!))
    productRepository.delete(product)
    assertNull(productRepository.find(product.sku!!))
  }

  @Test
  fun `Must throw an exception when try to update a product with an invalid sku`() {
    val productRepository: ProductRepository = ProductRepository()
    val warehouse = Warehouse("PR", 100, WarehouseType.ECOMMERCE)
    val inventory = Inventory(mutableListOf(warehouse))
    val product = Product(sku = 112233, name = "Product 9", inventory = inventory)

    assertThrows(NotFoundException::class.java, Executable { productRepository.update(product) })
  }

  @Test
  fun `Must throw an exception when try to delete a product with an invalid sku`() {
    val productRepository: ProductRepository = ProductRepository()
    val warehouse = Warehouse("PR", 100, WarehouseType.ECOMMERCE)
    val inventory = Inventory(mutableListOf(warehouse))
    val product = Product(sku = 112233, name = "Product 9", inventory = inventory)

    assertThrows(NotFoundException::class.java, Executable { productRepository.delete(product) })
  }

}
